02-peer-protocol.md

```diff
@@ -50,10 +67,451 @@ pubkey corresponding to the funding output nothing prevents duplicative channel
+
+Either node:
+  - MAY omit this message
+
+The sending node:
+  - MUST add all sent outputs to the collaborative transaction
+  - MUST ensure the `script` is a standard script
+  - MUST use a unique `serial_id` for each output
+  - MUST specify a `sats` value greater than the dust limit
+
+The receiving node:
+  - MUST add all received outputs to the collaborative transaction
+  - MUST fail the transaction collaboration if:
+    - it receives a duplicate `serial_id`
+    - it receives a `serial_id` from the peer with the incorrect parity
+
+
+#### Rationale
+Each node must have a complete set of the transaction outputs
+to derive the funding transaction and subsequent commitment signatures.
+
+`serial_id` is a randomly chosen number which uniquely identifies this output.
+Outputs in the final transaction will be sorted by `serial_id`.
+
+`sats` is the satoshi value of the output.
+
+`output_info`.`script` is the scriptPubKey for the output.
+
+
+### The `tx_remove_input` Message
+This message removes an input from the transaction
+
+1. type: 68 (`tx_remove_input`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u16`:`serial_id`]
+
+#### Requirements
+
+The sending node:
+  - MUST NOT send a `tx_remove_input` for an input which is not theirs
+
+The receiving node:
+  - MUST remove the indicated input from the final transaction
+  - MAY ignore the message if:
+    - the `serial_id` is novel (has not been communicated previously)
+  - MUST fail the channel if:
+    - the `serial_id` matches one of the receiving node's inputs
+
+
+### The `tx_remove_output` Message
+This message removes an output from the transaction
+
+1. type: 69 (`tx_remove_output`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u16`:`serial_id`]
+
+#### Requirements
+
+The sending node:
+  - MUST NOT send a `tx_remove_output` for an input which is not theirs
+
+The receiving node:
+  - MUST remove the indicated output from the final transaction
+  - MAY ignore the message if:
+    - the `serial_id` is novel (has not been communicated previously)
+  - MUST fail the channel if:
+    - the `serial_id` matches one of the receiving node's outputs
+
+
+### The `tx_complete` Message
+
+This message signals the conclusion of a peer's transaction
+contributions.
+
+1. type: 70 (`tx_complete`)
+2. data:
+    * [`channel_id`:`channel_id`]
+
+
+#### Requirements
+
+Both nodes:
+  - MUST send this message in succession for the collaboration to be considered concluded
+
+The receiving node:
+  - SHOULD use the provided inputs and outputs to construct a transaction
+  - MUST fail the channel if:
+    - any `serial_id` is duplicated
+    - the total satoshis of the inputs is less than the outputs
+    - any output value is less than the `dust_limit`.
+    - the peer's paid feerate does not meet or exceed the agreed `feerate`
+    - the `initiator`'s fees do not cover the `common` fields
+    - any input is duplicated (same txid:index)
+    - any input is non-Segwit
```

Comment:
I'm a little confused as to why there are requirements on `serial_id`, `dust_limit` and inputs when none of these things should be possible to be added to the tx. Though I suppose extra validation can't hurt

Reply:
> why there are requirements on serial_id

duplicate `serial_id`s would result in an undefined sort order, potentially causing a mismatch in resulting txs; ditto on having a different idea of what is 'dust_limit' compatible as in theory you would trim any sub-dust_limit outputs.